apps/src/code-studio/stageLockRedux.js

```diff
@@ -0,0 +1,240 @@
+/**
+ * Reducer and actions for stage lock info. This includes the teacher panel on
+ * the course overview page, and the stage locking dialog.
+ */
+
+import _ from 'lodash';
+import { makeEnum } from '@cdo/apps/utils';
+
+export const ViewType = makeEnum('Student', 'Teacher');
+export const LockStatus = makeEnum('Locked', 'Editable', 'Readonly');
+
+// Action types
+const SET_VIEW_TYPE = 'stageLock/SET_VIEW_TYPE';
+const SET_SECTIONS = 'stageLock/SET_SECTIONS';
+const SELECT_SECTION = 'stageLock/SELECT_SECTION';
+const OPEN_LOCK_DIALOG = 'stageLock/OPEN_LOCK_DIALOG';
+const CLOSE_LOCK_DIALOG = 'stageLock/CLOSE_LOCK_DIALOG';
+const BEGIN_SAVE = 'stageLock/BEGIN_SAVE';
+const FINISH_SAVE = 'stageLock/FINISH_SAVE';
+
+const initialState = {
+  viewAs: ViewType.Teacher,
+  sections: {},
+  selectedSection: null,
+  sectionsLoaded: false,
+  unlockedStageIds: [],
+  lockDialogStageId: null,
+  // The locking info for the currently selected section/stage
+  lockStatus: [],
+  saving: false
+};
+
+/**
+ * Stage lock reducer
+ */
+export default function reducer(state = initialState, action) {
+  if (action.type === SET_VIEW_TYPE) {
+    return Object.assign({}, state, {
+      viewAs: action.viewAs
+    });
+  }
+
+  if (action.type === SET_SECTIONS) {
+    const sectionId = Object.keys(action.sections)[0];
+    const currentSection = action.sections[sectionId];
+    return Object.assign({}, state, {
+      sections: action.sections,
+      sectionsLoaded: true,
+      selectedSection: sectionId,
+      unlockedStageIds: unlockedStages(currentSection)
```

Comment:
It looks like this information can be derived from other information on the state.  Is there a reason we're storing this instead of calculating it as-needed?